sys/extimer/extimer.c

```diff
@@ -0,0 +1,183 @@
+        DEBUG("%s():%d: integer overflow, returning UINT64_MAX\n", DEBUG_FUNC,
+              __LINE__);
+        return UINT64_MAX;
+    }
+    DEBUG("%s():%d: returning %" PRIu64 "\n", DEBUG_FUNC, __LINE__,
+          now + offset);
+    return now + offset;
+}
+
+void extimer_add(extimer_t *timer, extimer_event_t *event, kernel_pid_t pid)
+{
+    DEBUG("%s(timer=%p, event=%p, pid=%" PRIkernel_pid ")\n", DEBUG_FUNC,
+          (void *)timer, (void *)event, pid);
+    assert(timer);
+    assert(event->next == NULL);
+    assert(pid != KERNEL_PID_UNDEF);
+    DEBUG("%s():%d: entering mutex[%p]\n", DEBUG_FUNC, __LINE__,
+          (void *)&timer->mutex);
+    mutex_lock(&timer->mutex);
+    event->msg.sender_pid = pid;    /* use msg_t::sender_pid to store receiver
+                                     * until actual sending */
+    if (timer->events == NULL) {
+        DEBUG("%s():%d: no events in current timer, schedule immediately at %"
+              PRIu64 " usec\n", DEBUG_FUNC, __LINE__, event->time);
+        timer->events = event;
+        _time_next_event(timer);
+    }
+    else {
+        extimer_event_t *ptr = NULL, *tmp = timer->events;
+        while ((tmp != NULL) && (tmp->time <= event->time)) {
+            if (event == tmp) {   /* check rest of the list for identities to event */
+                DEBUG("%s():%d: event already scheduled. Doing nothing\n",
+                      DEBUG_FUNC, __LINE__);
+                mutex_unlock(&timer->mutex);
+                DEBUG("%s():%d: exiting mutex[%p]\n", DEBUG_FUNC, __LINE__,
+                      (void *)&timer->mutex);
+                return;
+            }
+            ptr = tmp;
+            tmp = tmp->next;
+        }
+        if (ptr == NULL) {  /* replace head */
+            DEBUG("%s():%d: schedule immediately at %" PRIu64 " usec\n",
+                  DEBUG_FUNC, __LINE__, event->time);
+            event->next = timer->events;
+            timer->events = event;
+            _time_next_event(timer);
+        }
+        else {
+            DEBUG("%s():%d: schedule later\n", DEBUG_FUNC, __LINE__);
+            list_add((list_node_t *)ptr, (list_node_t *)event);
+        }
+    }
+    mutex_unlock(&timer->mutex);
+    DEBUG("%s():%d: exiting mutex[%p]\n", DEBUG_FUNC, __LINE__,
+          (void *)&timer->mutex);
+}
+
+extimer_event_t *extimer_del(extimer_t *timer, extimer_event_t *event)
+{
+    extimer_event_t *res = NULL;
+    unsigned state;
+
+    DEBUG("%s(timer=%p, event=%p)\n", DEBUG_FUNC, (void *)timer, (void *)event);
+    assert(timer != NULL);
+    assert(event != NULL);
+
+    DEBUG("%s():%d: entering mutex[%p]\n", DEBUG_FUNC, __LINE__,
+          (void *)&timer->mutex);
+    mutex_lock(&timer->mutex);      /* lock mutex to wait for threads that might
+                                     * still use the timer */
+    DEBUG("%s():%d: disabling IRQ\n", DEBUG_FUNC, __LINE__);
+    state = irq_disable();          /* disable IRQ to not fire timer during
```

Comment:
But `xtimer_remove()` is only called when the first event in line is removed. And even then the next event after that could come in an interval below the threshold of xtimer... I'd like to prevent such race conditions from the start.